pkg/csi/service/vanilla/controller.go

```diff
@@ -1796,42 +1790,25 @@ func (c *controller) createFileVolume(ctx context.Context, req *csi.CreateVolume
 						scParams.StoragePolicyName, vcenter.Config.Host)
 				}
 
-				var candidateDatastores []*cnsvsphere.DatastoreInfo
-				// Get accessible datastores in the matching topology requirement.
-
-				vcenter, err = c.managers.VcenterManager.GetVirtualCenter(ctx, vcHost)
+				// Get accessible datastores in the VC's topology requirement.
+				candidateDatastores, err := placementengine.GetAllAccessibleDSInTopology(ctx,
+					topologySegmentsList, vcenter)
 				if err != nil {
-					return nil, faultType, logger.LogNewErrorCodef(log, codes.Internal,
-						"failed to get vCenter. Error: %+v", err)
-				}
-				if topologyRequirement.GetPreferred() != nil {
-					log.Debugf("Using preferred topology")
-					candidateDatastores, err = getCandidateDSInTopologyForFileVolumes(ctx,
-						topologyRequirement.GetPreferred(), vcenter)
-					if err != nil {
-						log.Errorf("Error finding candidate datastores using preferred topology: %+v",
-							topologyRequirement.GetPreferred())
-						return nil, csifault.CSIInternalFault, err
-					}
-				}
-				// If there are no datastores for the preferred topology requirement, fetch shared
-				// datastores for the requisite topology requirement instead.
-				if len(candidateDatastores) == 0 && topologyRequirement.GetRequisite() != nil {
-					log.Debugf("Using requisite topology")
-					candidateDatastores, err = getCandidateDSInTopologyForFileVolumes(ctx,
-						topologyRequirement.GetRequisite(), vcenter)
-					if err != nil {
-						log.Errorf("Error finding shared datastores using requisite topology: %+v",
-							topologyRequirement.GetRequisite())
-						return nil, csifault.CSIInternalFault, err
-					}
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"error finding candidate datastores in topology %+v for VC: %q",
+						topologySegmentsList, vcHost)
 				}
 				if len(candidateDatastores) == 0 {
-					log.Infof("No candidate datastores found for vcenter %q", vcenter.Config.Host)
+					errMsg := fmt.Sprintf("No candidate datastores available for file volume provisioning "+
+						"in topology: %+v in VC: %q", topologySegmentsList, vcenter.Config.Host)
+					log.Warn(errMsg)
+					combinedErrMssgs = append(combinedErrMssgs, errMsg)
 					continue
 				}
 
-				fsEnabledClusterToDsInfoMap = c.authMgrs[vcHost].GetFsEnabledClusterToDsMap(ctx)
+				// Filter File service enabled DS from candidate datastores.
+				var fsEnabledCandidateDatastores []*cnsvsphere.DatastoreInfo
+				fsEnabledClusterToDsInfoMap := c.authMgrs[vcHost].GetFsEnabledClusterToDsMap(ctx)
```

Comment:
when multi VC FSS is not ebaled, should we get fs enbaled datastore like this: fsEnabledClusterToDsInfoMap = c.authMgr.GetFsEnabledClusterToDsMap(ctx)

Reply:
The code protected by the new FSS `topology-aware-file-volume` mostly assumes that the older FSS are set to true. If we choose to honor both the FSS simultaneously, the code will get very complex.